pkg/provider/docker/docker.go

```diff
@@ -108,46 +109,75 @@ type networkData struct {
 }
 
 func (p *Provider) createClient() (client.APIClient, error) {
-	var httpClient *http.Client
+	opts, err := p.getClientOpts()
+	if err != nil {
+		return nil, err
+	}
+
+	httpHeaders := map[string]string{
+		"User-Agent": "Traefik " + version.Version,
+	}
+	opts = append(opts, client.WithHTTPHeaders(httpHeaders))
+
+	apiVersion := DockerAPIVersion
+	if p.SwarmMode {
+		apiVersion = SwarmAPIVersion
+	}
+	opts = append(opts, client.WithVersion(apiVersion))
+
+	return client.NewClientWithOpts(opts...)
+}
+
+func (p *Provider) getClientOpts() ([]client.Opt, error) {
+	helper, err := connhelper.GetConnectionHelper(p.Endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	// SSH
+	if helper != nil {
+		httpClient := &http.Client{
+			Transport: &http.Transport{
+				DialContext: helper.Dialer,
+			},
+		}
+
+		return []client.Opt{
+			client.WithHTTPClient(httpClient),
+			client.WithHost(helper.Host), // To avoid 400 Bad Request: malformed Host header daemon error
+			client.WithDialContext(helper.Dialer),
+		}, nil
```

Comment:
as on `client.WithHTTPClient` you only change the `DialContext`, you can use `client.WithDialContext` only.

Reply:
Really? I thought `WithHTTPClient` says that Docker should connect via given HTTP